### ISSUE TITLE
Messages out of order between cluster members

### ISSUE THREAD

**Issue by umeat** [emqx/emqx#4971]
**Environment**:

- EMQ X version (e.g. `emqx_ctl status`): 4.2.9 
- Hardware configuration (e.g. `lscpu`): Kubernetes EKS and K3S

**What happened and what you expected to happen**:

Running a cluster in Kubernetes with 3 pods as a statefulset I get out of order messages depending on which cluster member I subscribe to.

**How to reproduce it (as minimally and precisely as possible)**:

(pods are emqx-0, 1, and 2) Publish a message to emqx-0 on topic "foo" every millisecond. If I subscribe to emqx-0 all messages are received by the subscribing client in order. However, if I subscribe on emqx-1 or emqx-2, then the messages are sometimes randomly out of order.

This can happen with much lower message rates if there is more traffic on the nodes, but I can reproduce it locally with the above and no other traffic.

**Comment by Rory-Z** [comment]
Hi，@umeat
Can you provide a detailed log for emqx?

**Comment by umeat** [comment]
There is nothing logged about it. Anything in particular I can check for?

**Comment by Rory-Z** [comment]
> There is nothing logged about it. Anything in particular I can check for?

You can set log level to debug

**Comment by umeat** [comment]
Yeah, with debug logging I don't notice anything unusual, but you can see here the messages coming out of order on emqx-1 (where the subscriber is connected) and not on emqx-2 (where the publisher is connected) - the message payload is a number left padded with 0's:

```
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.354 [debug] <<"CLIENT">>@127.0.0.1:33598 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=rtcm3/binary/foo/771, PacketId=undefined, Payload=<<"00000000000000000000000000000000000000000000506750">>)
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.356 [debug] event=cast_received driver=tcp socket="#Port<0.608>" peer="10.42.0.11:33389" module=emqx_broker function=dispatch args="[<<"rtcm3/binary/foo/#">>,{delivery,<42654.27478.15>,{message,<<0,5,196,104,36,216,92,221,159,101,15,0,107,86,187,129>>,1,<<"426c06b4-ff2f-46d4-81ba-d43c7f051820">>,#{dup => false,retain => false},#{peerhost => {10,42,0,18},properties => #{},proto_ver => 4,protocol => mqtt,username => <<"test">>},<<"rtcm3/binary/foo/771">>,<<"00000000000000000000000000000000000000000000506752">>,1623326457355}}]"
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.356 [debug] <<"CLIENT">>@127.0.0.1:33598 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=rtcm3/binary/foo/771, PacketId=undefined, Payload=<<"00000000000000000000000000000000000000000000506752">>)
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.356 [debug] event=cast_received driver=tcp socket="#Port<0.608>" peer="10.42.0.11:33389" module=emqx_broker function=dispatch args="[<<"rtcm3/binary/foo/#">>,{delivery,<42654.27478.15>,{message,<<0,5,196,104,36,216,88,146,159,101,15,0,107,86,187,128>>,1,<<"426c06b4-ff2f-46d4-81ba-d43c7f051820">>,#{dup => false,retain => false},#{peerhost => {10,42,0,18},properties => #{},proto_ver => 4,protocol => mqtt,username => <<"test">>},<<"rtcm3/binary/foo/771">>,<<"00000000000000000000000000000000000000000000506751">>,1623326457354}}]"
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.357 [debug] <<"CLIENT">>@127.0.0.1:33598 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=rtcm3/binary/foo/771, PacketId=undefined, Payload=<<"00000000000000000000000000000000000000000000506751">>)
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.357 [debug] event=cast_received driver=tcp socket="#Port<0.608>" peer="10.42.0.11:33389" module=emqx_broker function=dispatch args="[<<"rtcm3/binary/foo/#">>,{delivery,<42654.27478.15>,{message,<<0,5,196,104,36,216,97,83,159,101,15,0,107,86,187,130>>,1,<<"426c06b4-ff2f-46d4-81ba-d43c7f051820">>,#{dup => false,retain => false},#{peerhost => {10,42,0,18},properties => #{},proto_ver => 4,protocol => mqtt,username => <<"test">>},<<"rtcm3/binary/foo/771">>,<<"00000000000000000000000000000000000000000000506753">>,1623326457356}}]"
(emqx@emqx-1.emqx-headless.gnss-stream.svc.cluster.local)1> 2021-06-10 12:00:57.357 [debug] <<"CLIENT">>@127.0.0.1:33598 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=rtcm3/binary/foo/771, PacketId=undefined, Payload=<<"00000000000000000000000000000000000000000000506753">>)
```

You can see 506750 followed by 506752 followed by 506751 followed by 506753.

Whereas my client subscribing on emqx-2 receives the messages in order.

**Comment by ertanden** [comment]
We are facing very similar issue with EMQ-X 4.3.2 on Kubernetes with 3 nodes. Seems like cross-node order is not preserved for the same topic and publisher/subscriber pair.

**Comment by k32** [comment]
Disclaimer: I am not very familiar with the EMQX router code. It looks like it forward messages in sync and async modes depending on the value of `rpc.mode` config parameter. By default, `async` is used. Can you check if you set the value of this parameter?

In async mode `gen_rpc:cast` is used, and I am pretty sure it changes ordering.

**Comment by zmstone** [comment]
I remember similar issue was fixed by setting this config: `rpc.tcp_client_num=1`

EMQX by default set up `NumberOfSchedulers / 2` connections towards each peer node.
We should change default to `1` before we have a consistent dispatching mechanism added to the rpc library.

**Comment by umeat** [comment]
Setting both of those `rpc` variables resolved this issue. Thanks very much.

Is there any obvious downside to these settings? Performance hits for example?

**Comment by zmstone** [comment]
I am not aware if there were performance tests done for the configuration tuning.
Gut feeling is it's ok to use single TCP connection in common scenarios.
@terry-xiaoyu maybe knows more.

**Comment by zmstone** [comment]
PR to change the default value of this config:
https://github.com/emqx/emqx/pull/4979

**Comment by zmstone** [comment]
turns out the fix #4979 was wrong.
will have a proper fix in v4.4 https://github.com/emqx/emqx/pull/6024
